onionshare_gui/main_window.py

```diff
@@ -0,0 +1,292 @@
+            settings_button = QtWidgets.QPushButton(
+                strings._("gui_tor_connection_ask_open_settings")
+            )
+            quit_button = QtWidgets.QPushButton(
+                strings._("gui_tor_connection_ask_quit")
+            )
+            a.addButton(settings_button, QtWidgets.QMessageBox.AcceptRole)
+            a.addButton(quit_button, QtWidgets.QMessageBox.RejectRole)
+            a.setDefaultButton(settings_button)
+            a.exec_()
+
+            if a.clickedButton() == settings_button:
+                # Open settings
+                self.common.log(
+                    "OnionShareGui",
+                    "_tor_connection_canceled",
+                    "Settings button clicked",
+                )
+                self.open_settings()
+
+            if a.clickedButton() == quit_button:
+                # Quit
+                self.common.log(
+                    "OnionShareGui", "_tor_connection_canceled", "Quit button clicked"
+                )
+
+                # Wait 1ms for the event loop to finish, then quit
+                QtCore.QTimer.singleShot(1, self.common.gui.qtapp.quit)
+
+        # Wait 100ms before asking
+        QtCore.QTimer.singleShot(100, ask)
+
+    def tor_connection_open_settings(self):
+        """
+        The TorConnectionDialog wants to open the Settings dialog
+        """
+        self.common.log("MainWindow", "tor_connection_open_settings")
+
+        # Wait 1ms for the event loop to finish closing the TorConnectionDialog
+        QtCore.QTimer.singleShot(1, self.open_settings)
+
+    def open_settings(self):
+        """
+        Open the SettingsDialog.
+        """
+        self.common.log("MainWindow", "open_settings")
+        d = SettingsDialog(self.common)
+        d.settings_saved.connect(self.settings_have_changed)
+        d.exec_()
+
+    def settings_have_changed(self):
+        self.common.log("OnionShareGui", "settings_have_changed")
+
+        if self.common.gui.onion.is_authenticated():
+            self.status_bar.clearMessage()
+
+        # Tell each tab that settings have changed
+        for index in range(self.tabs.count()):
+            tab = self.tabs.widget(index)
+            tab.settings_have_changed()
+
+    def check_for_updates(self):
+        """
+        Check for updates in a new thread, if enabled.
+        """
+        if self.common.platform == "Windows" or self.common.platform == "Darwin":
+            if self.common.settings.get("use_autoupdate"):
+
+                def update_available(update_url, installed_version, latest_version):
+                    Alert(
+                        self.common,
+                        strings._("update_available").format(
+                            update_url, installed_version, latest_version
+                        ),
+                    )
+
+                self.update_thread = UpdateThread(
+                    self.common, self.common.gui.onion, self.common.gui.config
```

Comment:
`self.common.gui.config` is an unknown attribute and crashes Mac (probably Windows too)